src/librustc_typeck/mem_categorization.rs

```diff
@@ -679,20 +762,67 @@ impl<'a, 'tcx> MemCategorizationContext<'a, 'tcx> {
         op(&place_with_id, pat);
 
         match pat.kind {
-            PatKind::TupleStruct(_, ref subpats, _) | PatKind::Tuple(ref subpats, _) => {
-                // S(p1, ..., pN) or (p1, ..., pN)
+            PatKind::Tuple(ref subpats, _) => {
+                // (p1, ..., pN)
                 for subpat in subpats.iter() {
                     let subpat_ty = self.pat_ty_adjusted(&subpat)?;
-                    let sub_place = self.cat_projection(pat, place_with_id.clone(), subpat_ty);
+                    let sub_place = self.cat_projection(
+                        pat,
+                        place_with_id.clone(),
+                        subpat_ty,
+                        ProjectionKind::Other,
```

Comment:
Hmm, I think these should be treated as fields as well, with variant 0 and indices 0...N.